src/python/pants/core/util_rules/external_tool_test.py

```diff
@@ -105,3 +108,66 @@ def do_test(
         create_subsystem(
             FooBar, version="9.9.9", known_versions=FooBar.default_known_versions
         ).get_request(Platform.darwin)
+
+
+class ConstrainedTool(TemplatedExternalTool):
+    name = "foobar"
+    options_scope = "foobar"
+    version_constraints = ">3.2.1, <3.8"
+    default_version = "3.4.7"
+    default_known_versions = [
+        "3.2.0|darwin   |1102324cdaacd589e50b8b7770595f220f54e18a1d76ee3c445198f80ab865b8|123346",
+        "3.2.0|linux_ppc|39e5d64b0f31117c94651c880d0a776159e49eab42b2066219569934b936a5e7|124443",
+        "3.2.0|linux    |c0c667fb679a8221bed01bffeed1f80727c6c7827d0cbd8f162195efb12df9e4|121212",
+        "3.4.7|darwin   |9d0e18cd74b918c7b3edd0203e75569e0c8caecb1367b3be409b45e28514f5be|123321",
+        "3.4.7|linux    |a019dfc4b32d63c1392aa264aed2253c1e0c2fb09216f8e2cc269bbfb8bb49b5|134213",
+    ]
+    default_url_template = "https://foobar.org/bin/v{version}/foobar-{version}-{platform}.tgz"
+    default_url_platform_mapping = {
+        "darwin": "osx-x86_64",
+        "linux": "linux-x86_64",
+    }
+
+    def generate_exe(self, plat: Platform) -> str:
+        return f"foobar-{self.version}/bin/foobar"
+
+
+@contextmanager
+def no_exception():
+    yield None
```

Comment:
hmm.. I've used this construct twice now, in two different PR's.. maybe move these to a test util module, unless you oppose to the testing method..

Reply:
I love declarative tests, so: by all means! Maybe a new pytest-specific module in `src/python/pants/testutil`?